### ISSUE TITLE
Replicating the DINOv2 Experiment from Table 3

### ISSUE THREAD

**Issue by Saloni1Parekh609** [NVlabs/RADIO#99]
I'm trying to replicate the DINOv2 experiment from this table:
![Testing DINOv2 Distillation](https://github.com/user-attachments/assets/67353690-fb39-454e-8975-f1458c1bf2c8)

However, I had some questions here. 
1. Is this student model trained with a 224x224 resolution with CPE? Does DINOv2 sees a 224x224 resolution image as well? For evaluation on ADE20k, is the input data of 512x512 resolution? And is it 224x224 for the CLIP distilled model?
2. What are the learning rate hyperparameters set for this experiment? Is it the same as the RADIO experiments? How many steps did you run this for?
3. Could you provide a little more clarity behind the Learning Rate Scheduler for the RADIO experiment? For example, the cycling aspect of CosineAnnealing scheduler.

**Comment by Saloni1Parekh609** [comment]
I had another quick question for you. To clarify, does this table mean that for row 2 for example, you distilled SAM and DINOv2 **without** clip and then evaluated the model? So SAM is always a teacher for these experiments?

On another note, for zero shot evaluation do you also use the adaptor heads? Or do you only use the encoder's representations before they are fed into the adaptor?

**Comment by mranzinger** [comment]
Actually, for Table 3, we didn't use SAM at all. In general, we struggled with SAM a lot until subsequent works, such as PHI-S and also an upcoming paper.

For zero shot, we use the adaptor head for CLIP, which is necessary because the summary adaptor output is what's aligned with CLIP's text encoder.

For kNN, we use the base model features, so no adaptors.

**Comment by Saloni1Parekh609** [comment]
i see, so what would none mean in this table?

**Comment by mranzinger** [comment]
That table is which models, if any, we're supervising the spatial features. So "none" would mean that we're still applying the summary losses, but not the spatial feature losses.

**Comment by Saloni1Parekh609** [comment]
ah i see, got it thanks!
and thaks for the prompt responses as well!

**Comment by mranzinger** [comment]
Happy to help. What we noticed was that DINOv2 had much better (cleaner?) spatial features than CLIP, so we wanted to know if also distilling CLIP's features helped or hurt the student. For completeness, we studied CLIP-only, and also none, which in total tells us that the student benefits from learning the features of both teachers. We didn't include SAM in that ablation, but I have more recent results that suggest that (when you properly balance the teachers, see the PHI-S paper), it's beneficial to also distill SAM's features. We'll be able to share those results hopefully soon.

**Comment by mranzinger** [comment]
I've tried it both ways, and found the difference to be negligible, owing probably to AdamW's relative robustness to changes in learning rate over a reasonable window.

**Comment by Saloni1Parekh609** [comment]
Thank you!

**Comment by Saloni1Parekh609** [comment]
@mranzinger  I've been trying to recreate this experiment but my training doesnt seem to converge. 

From table 2, if I was using METACLIP/H14 and DINOv2/G14 and distilling spatial and summary features to a VITL14 student model, my initial lr is 1e-5, warm up is 1000 steps to get to 1e-3 and then a cosine annealing decay to 0. I am using ADAMW, my loss keeps going up. The input resolution for both teachers  and student is 224 x 224.

Is there something I am doing incorrectly here?

**Comment by mranzinger** [comment]
What's your batch size? Is the run always divergent, or just after some number of steps?

**Comment by Saloni1Parekh609** [comment]
Its 128 on 8 GPUs each for CLIP and DINO only. Its always divergent, as soon as the training starts.

**Comment by mranzinger** [comment]
Just to make sure I understand, the batch size is 128 for each GPU, resulting in a total size of 1024, right? Are you using the 2-layer MLP formulation for the adaptor heads? Are you using a different adaptor head for each teacher's summary and spatial features? (e.g. if you have 2 teachers, you should have 4 adaptors)

Immediate divergence strongly suggests a bug somewhere in your code. Couple things you can try to sleuth the problem out:
1) Use a dataset that's a single image, without augmentation. We call this an overfit test. If your model can't memorize the expected outputs, then something is wrong somewhere.
2) Try initializing your student from one of the teachers. Presumably the matching loss for the student/teacher pair should be quite small. It might help to switch to an identity initialized linear layer as your adaptor head, as then there should literally be no difference between the student and one of the teachers before the first backprop. An ideal training curve here should be a pretty rapid drop in matching loss for the other teacher, and a (hopefully mild) increase in loss for the copied teacher.

**Comment by Saloni1Parekh609** [comment]
hm, with a single image, the loss goes up after thr lr reaches 1e-3. i was able to make it work with a lower peak lr (5e-4)

**Comment by mranzinger** [comment]
What value for weight decay are you using? You very well may be dealing with [attention entropy collapse](https://proceedings.mlr.press/v202/zhai23a.html) albeit we didn't really run into that issue until we tried hi-res training with SAM.

**Comment by Saloni1Parekh609** [comment]
It is 0.02 :/

**Comment by Saloni1Parekh609** [comment]
I see, looking forward to it!

With a single teacher, it explodes. The same story, I could get it to work with a lower lr.

**Comment by mranzinger** [comment]
Yeah, there's definitely a bug in your setup somewhere then, unfortunately.

**Comment by Saloni1Parekh609** [comment]
was your precision float 32 for amradio (pre phi)? was the feature normalisation necessary for bfloat 16?

**Comment by mranzinger** [comment]
We train using AMP (Automatic mixed precision) with the `torch.bfloat16` dtype. No grad scaler.
